src/KubernetesClient.Informers/ResourceObservableExtensions.cs

```diff
@@ -0,0 +1,273 @@
+                            observer.OnNext(itemSelector(item));
+                        resetBuffer.Clear();
+                    }
+
+                    void OnComplete()
+                    {
+                        FlushBuffer();
+                        observer.OnCompleted();
+                    }
+
+                    void OnError(Exception e)
+                    {
+                        observer.OnError(e);
+                    }
+
+                    var upstreamSubscription = source
+                        .Subscribe(notification =>
+                        {
+                            if (notification.EventFlags.HasFlag(EventTypeFlags.Reset))
+                            {
+                                resetBuffer.Add(notification);
+                                if (!notification.EventFlags.HasFlag(EventTypeFlags.ResetEnd)) // continue buffering till we reach the end of list window
+                                    return;
+                            }
+
+                            if (notification.EventFlags.HasFlag(EventTypeFlags.ResetEnd))
+                            {
+                                FlushBuffer();
+                                return;
+                            }
+
+                            if (!notification.EventFlags.HasFlag(EventTypeFlags.Reset) && resetBuffer.Count > 0) FlushBuffer();
+
+                            observer.OnNext(itemSelector(notification));
+                        }, OnError, OnComplete);
+                    return StableCompositeDisposable.Create(upstreamSubscription, Disposable.Create(OnComplete));
+                })
+                .ObserveOn(Scheduler.Immediate);
+        }
+
+        /// <summary>
+        ///     Synchronizes the specified cache with resource event stream such that cache is maintained up to date.
+        /// </summary>
+        /// <param name="source">The source sequence</param>
+        /// <param name="cache">The cache to synchronize</param>
+        /// <param name="keySelector">The key selector function</param>
+        /// <typeparam name="TKey">The type of key</typeparam>
+        /// <typeparam name="TResource">The type of resource</typeparam>
+        /// <returns>Source sequence wrapped into <see cref="CacheSynchronized{T}" />, which allows downstream consumers to synchronize themselves with cache version</returns>
+        public static IObservable<CacheSynchronized<ResourceEvent<TResource>>> SynchronizeCache<TKey, TResource>(
+            this IObservable<ResourceEvent<TResource>> source,
+            ICache<TKey, TResource> cache,
+            Func<TResource, TKey> keySelector,
+            Func<TResource, TResource, TResource> mapper = null)
+        {
+            if (mapper == null)
+                mapper = (oldResource, newResource) => newResource;
+
+            return Observable.Defer(() =>
+            {
+                long msgNum = 0;
+
+                return source
+                    .Do(_ => msgNum++)
```

Comment:
This `msgNum` being stored outside but mutated with a Do within the observable concerns me. With the purpose of it being simply a Message Number, I expect it's probably fine... but I can't help but think there's a better way.

Reply:
As long as closure semantics are well understood, this is safe to do. It's done in a few other places. RX itself uses a special pattern that is based on stateful classes to model new RX operators, but all this infrastructure is internal. I wanted to keep this light.

Reply:
Thanks for the explanation - it makes sense with this being in a new operator.